src/DictVectors/dvec.jl

```diff
@@ -1,113 +1,156 @@
-DVec(args...; capacity) = DVec(Dict(args...), capacity)
+Dictionary-based vector-like data structure for use with FCIQMC and
+[KrylovKit](https://github.com/Jutho/KrylovKit.jl). While mostly behaving like a `Dict`, it
+supports various linear algebra operations such as `norm` and `dot`. It has a
+[`StochasticStyle`](@ref) that is used to select an appropriate spawning strategy in the
+FCIQMC algorithm.
 
-function DVec(dict::D, capacity::Int) where D <: Dict
-    sizehint!(dict, capacity)
-    return DVec(dict)
-end
+See also: [`AbstractDVec`](@ref).
+
+# Constructors
+
+* `DVec(dict::AbstractDict[, style, capacity])`: create a `DVec` with `dict` for storage.
+  Note that the data may or may not be copied.
+
+* `DVec(args...[; style, capacity])`: `args...` are passed to the `Dict` constructor. The
+  `Dict` is used for storage.
+
+* `DVec{K,V}([style, capacity])`: create an empty `DVec{K,V}`.
+
+* `DVec(dv::AbstractDVec[, style, capacity])`: create a `DVec` with the same contents as
+   `adv`. The `style` is inherited from `dv` by default.
 
-# by specifying keytype, eltype, and capacity
-DVec{K,V}(capacity::Int) where V where K = DVec(Dict{K,V}(), capacity)
+The default `style` is selected based on the `DVec`'s `valtype` (see
+[`default_style`](@ref)). If a style is given and the `valtype` does not match the `style`'s
+`eltype`, the values are converted to an appropriate type.
 
-# from Vector
-function DVec(v::AbstractVector{T}, capacity = length(v)) where T
-    indices = 1:length(v) # will be keys of dictionary
-    ps = map(tuple,indices,v) # create iterator over pairs
-    return DVec(Dict(ps), capacity)
+The capacity argument is optional and sets the initial size of the `DVec` via `sizehint!`.
+
+# Examples
+
+```jldoctest
+julia> dv = DVec(:a => 1)
+DVec{Symbol,Int64,IsStochasticInteger} with 1 entries
+  :a => 1
+
+julia> dv = DVec(:a => 2, :b => 3; style=IsDynamicSemistochastic())
+DVec{Symbol,Float32,IsDynamicSemistochastic{true}} with 2 entries
+  :a => 2.0f0
+  :b => 3.0f0
+```
+"""
+struct DVec{K,V,S<:StochasticStyle{V},D<:AbstractDict{K,V}} <: AbstractDVec{K,V}
+    storage::D
+    style::S
 end
 
-# from AbstractDict; note that a new dict is constructed and data is copied
-function DVec(d::AbstractDict{K,V},
-              capacity = length(d)) where K where V
-    dv = DVec{K,V}(capacity)
-    for (k,v) in d
-        dv[k] = v
+###
+### Constructors
+###
+function DVec(args::Vararg{Pair{K,V}}; style=default_style(V), capacity=0) where {K,V}
+    dict = Dict{K,V}()
+    sizehint!(dict, max(length(args), capacity))
+    for (k, v) in args
+        dict[k] = convert(eltype(style), v)
     end
-    return dv
+    return DVec(dict; style)
+end
+# In this constructor, the style matches the dict's valtype.
+function DVec(
+    dict::AbstractDict{K,V}; style::StochasticStyle{V}=default_style(V), capacity=0
+) where {K,V}
+    capacity > 0 && sizehint!(dict, capacity)
+    return DVec{K,V,typeof(style),typeof(dict)}(dict, style)
+end
```

Comment:
There seems to be a problem here, that I don't quite understand.
`DVec(SwissDict(:a=>3.0))` works but
```
julia> DVec(SwissDict(:a=>3.0f0))
ERROR: MethodError: no method matching var"#DVec#8"(::IsDeterministic, ::Int64, ::Type{DVec}, ::SwissDict{Symbol, Float32})
Closest candidates are:
  var"#DVec#8"(::StochasticStyle{V}, ::Any, ::Type{DVec}, ::AbstractDict{K, V}) where {K, V} at /Users/brand/git/Rimu/src/DictVectors/dvec.jl:61
Stacktrace:
 [1] DVec(dict::SwissDict{Symbol, Float32})
   @ Rimu.DictVectors ~/git/Rimu/src/DictVectors/dvec.jl:64
 [2] top-level scope
   @ none:1
```
while `DVec(Dict(:a=>3.0f0))` works again (using the next method) but converts `Float32` into `Float64`.

It is not a big deal at this point though. A bigger question is, maybe, whether we want to allow more flexible `StochasticStyle`s in the future that allowed arbitrary number types.